### PR TITLE
Fix bug XWALK-2498, Ssl Certification Error Alert

### PR DESCRIPTION
**Pull request by axinging**
Fix bug XWALK-2498, Ssl Certification Error Alert.

### CODE REVIEW

**Comment by wang16** [comment]
@axingxing, thanks for the quick fix. You probably need to look for the linux bot's failure. You are changing code not only affect Android port. Besides, you need to change your commit log to following formats.
```
[Android] title

message

BUG=XWALK-xxxx
```

**Comment by wang16** [comment]
@axinging , you'd better put what's the problem/how you fix it in the commit log instead of repeating the title.

**Comment by axinging** [comment]
@lincsoon @halton

File: runtime/browser/xwalk_browser_context.h [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include <map>
 #include <vector>
 
+#include "xwalk/runtime/browser/xwalk_ssl_host_state_delegate.h"
```
Comment (halton) [inline runtime/browser/xwalk_browser_context.h]: nit, order

File: runtime/browser/xwalk_browser_context.h [hunk 2]
```diff
@@ -116,7 +117,7 @@ class XWalkBrowserContext
       scoped_refptr<RuntimeURLRequestContextGetter> >
       PartitionPathContextGetterMap;
   PartitionPathContextGetterMap context_getters_;
-
+  scoped_ptr<XWalkSSLHostStateDelegate> ssl_host_state_delegate_;
```
Comment (halton) [inline runtime/browser/xwalk_browser_context.h]: nit, one empty line after

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+#include "xwalk/runtime/browser/xwalk_ssl_host_state_delegate.h"
+
+#include "net/base/hash_value.h"
+
+using content::SSLHostStateDelegate;
+
+namespace xwalk {
+
+namespace internal {
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]: nit, line break after

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+net::SHA256HashValue getChainFingerprint256(const net::X509Certificate& cert) {
+  net::SHA256HashValue fingerprint =
+      net::X509Certificate::CalculateChainFingerprint256(
+          cert.os_cert_handle(), cert.GetIntermediateCertificates());
+  return fingerprint;
+}
+
+CertPolicy::CertPolicy() {
+}
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]: nit, line break after

File: xwalk.gyp [hunk 5]
```diff
@@ -221,6 +221,8 @@
         'runtime/browser/xwalk_application_mac.mm',
         'runtime/browser/xwalk_browser_context.cc',
         'runtime/browser/xwalk_browser_context.h',
+        'runtime/browser/xwalk_ssl_host_state_delegate.cc',
+        'runtime/browser/xwalk_ssl_host_state_delegate.h',
```
Comment (wang16) [inline xwalk.gyp]: alphabetical order.

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+XWalkSSLHostStateDelegate::XWalkSSLHostStateDelegate() {
+}
+
+XWalkSSLHostStateDelegate::~XWalkSSLHostStateDelegate() {
+}
+
+void XWalkSSLHostStateDelegate::HostRanInsecureContent(const std::string& host,
+                                                    int pid) {
+  // Intentional no-op for Android WebView.
```
Comment (wang16) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]: Note that this file is not only for Android.

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+// For an allowance, we consider a given |cert| to be a match to a saved
+// allowed cert if the |error| is an exact match to or subset of the errors
+// in the saved CertStatus.
+bool CertPolicy::Check(const net::X509Certificate& cert,
+                       net::CertStatus error) const {
+  net::SHA256HashValue fingerprint = getChainFingerprint256(cert);
+  std::map<net::SHA256HashValue, net::CertStatus,
+           net::SHA256HashValueLessThan>::const_iterator allowed_iter =
+      allowed_.find(fingerprint);
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]: please use typedef in header to reduce the length.

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  if ((allowed_iter != allowed_.end()) && (allowed_iter->second & error) &&
+      ((allowed_iter->second & error) == error)) {
+    return true;
+  }
+  return false;
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]:
```
return (allowed_iter != allowed_.end() &&
       (allowed_iter->second & error) &&
       (allowed_iter->second & error) == error)
```

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+}  // namespace internal
+
+XWalkSSLHostStateDelegate::XWalkSSLHostStateDelegate() {
+}
+
+XWalkSSLHostStateDelegate::~XWalkSSLHostStateDelegate() {
+}
+
+void XWalkSSLHostStateDelegate::HostRanInsecureContent(const std::string& host,
+                                                    int pid) {
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]: 4-space indent or align parameters

File: runtime/browser/xwalk_ssl_host_state_delegate.cc [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+bool XWalkSSLHostStateDelegate::DidHostRunInsecureContent(
+    const std::string& host,
+    int pid) const {
+  // Intentional no-op for Android WebView.
+  return false;
+}
+
+void XWalkSSLHostStateDelegate::AllowCert(const std::string& host,
+                                       const net::X509Certificate& cert,
+                                       net::CertStatus error) {
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.cc]: wrong indent

File: runtime/browser/xwalk_ssl_host_state_delegate.h [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+class XWalkSSLHostStateDelegate : public content::SSLHostStateDelegate {
+ public:
+  XWalkSSLHostStateDelegate();
+  virtual ~XWalkSSLHostStateDelegate();
+
+  // Records that |cert| is permitted to be used for |host| in the future, for
+  // a specified |error| type.
+  void AllowCert(const std::string& host,
+                 const net::X509Certificate& cert,
+                 net::CertStatus error) override;
```
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.h]: OVERRIDE plese, ditto as below

File: xwalk.gyp (hunk 5 above)
Comment (halton) [inline xwalk.gyp]: order plz

File: runtime/browser/xwalk_ssl_host_state_delegate.h (hunk 11 above)
Comment (halton) [inline runtime/browser/xwalk_ssl_host_state_delegate.h]: Please ignore this comment, upstream recently remove this usage. https://code.google.com/p/chromium/issues/detail?id=417463

**Comment by halton** [comment]
LGTM with nit on git log: Ssl Certification -> SSL Certification

**Comment by halton** [comment]
LGTM

**Comment by lincsoon** [comment]
LGTM

**Comment by takethathe** [comment]
@axinging 
Seems that the patch impacts all platforms, have you ever tested on Tizen?

**Comment by wang16** [comment]
> @axinging 
> Seems that the patch impacts all platforms, have you ever tested on Tizen?

That's what i'm going to say. We don't know how to verify on tizen. But this code seems platform unrelated. It's better if it can serve Tizen as well. Good news is that at least bots passed.

@takethathe , do you have any suggestions that who can help do a simple verification? Or involve appropriate Tizen owner to review the patch as well.

**Comment by halton** [comment]
building for Tizen is not hard thing, and you can run testing on the [Tizen IVI vmware image] (https://wiki.tizen.org/wiki/IVI/IVI_3.0_VMware).

**Comment by axinging** [comment]
@wang16 @takethathe
After co-test with xinchao,  we found that tizen didn't implement the "Cancel"/"OK" ui for unknown SSL site. And when launch the test page, it just showed up a blank page. It behaves as patched before.

This patch has no influence on current tizen:
When launch test page at first time, it will query if there are accepted cert. Generally it return "DENIED". If "DENIED", it will  prompt  a  "Cancel"/"OK" ui. But currently tizen didn't implement this ui, so this patch has no influence on the behavior of tizen.

@axinging 
Seems that the patch impacts all platforms, have you ever tested on Tizen?
That's what i'm going to say. We don't know how to verify on tizen. But this code seems platform unrelated. It's better if it can serve Tizen as well. Good news is that at least bots passed.

@takethathe , do you have any suggestions that who can help do a simple verification? Or involve appropriate Tizen owner to review the patch as well.

**Comment by wang16** [comment]
LGTM then. Thanks for the patch.

**Pull request merged**